### PR TITLE
feat(gatsby-source-shopify): accept custom queries

### PR DESCRIPTION
**Pull request by rmcfadzean**
Allow users to pass custom queries to the gatsby-source-shopify plugin.

Created due to the need for [`totalInventory`](https://shopify.dev/docs/storefront-api/reference/object/product?api[version]=2020-04#totalinventory-2020-04) and [`quantityAvailable`](https://shopify.dev/docs/storefront-api/reference/object/productvariant?api[version]=2020-04#quantityavailable-2020-04) fields which are only accessible with the `unauthenticated_read_product_inventory` scope.

## Description

This PR includes a new config option for the gatsby-source-shopify plugin which allows users to use their own queries.

This can be particularly helpful for queries not normally 

#### New option `queries`

Defaults to the queries from `gatsby-source-shopify/src/queries.js`

Overrides must accept the arguments `first` & `after`.
eg:

```graphql
query GetProducts($first: Int!, $after: String) {
  products(first: $first, after: $after) {
   # fields go here
  }
}
```

### CODE REVIEW

**Comment by rmcfadzean** [comment]
Hi Dustin,

I gave it a try but unfortunately that's seems like there is *a lot* of work needed. The existing tests are very fragile and I'm at a loss on how to break down the scope. Any guidance here? I'd love to get better at writing tests.

As for your last point, yes. I am using this in a new site for a client (forked locally temporarily)

File: packages/gatsby-source-shopify/README.md [hunk 1]
```diff
@@ -70,6 +70,29 @@ plugins: [
       // Possible values are: 'shop' and 'content'.
       // Defaults to ['shop', 'content'].
       includeCollections: ["shop", "content"],
+
+      // Allow overriding the default queries
+      // This allows you to include/exclude extra fields when sourcing nodes
+      // Available keys are: articles, blogs, collections, products, shopPolicies, and pages
+      // Queries need to accept arguments for first and after
+      // Defaults can be found in ./src/queries.js
+      queries: {
```
Comment (smthomas) [inline packages/gatsby-source-shopify/README.md:79]: I think this config documentation needs to be updated to shopifyQueries in order for it to work.

**Comment by JGeiser9** [comment]
Just read through this PR and it would be a sweet addition since queries.js doesn't include 'totalInventory' by default. Out of curiosity, do the shopifyQueries ever actually get built if included in the plugin config?

**Comment by smthomas** [comment]
> Out of curiosity, where do the shopifyQueries actually get built when included in the plugin config?

I'm not sure if this answers your question, but the shopifyQueries config gets merged with the original queries from queries.js here - https://github.com/gatsbyjs/gatsby/pull/23922/files#diff-9f1a842cf44ff7772dcebf087a07eb00R70

**Comment by rmcfadzean** [comment]
Sorry about the delay @smthomas. I had just made the change seconds after you did!

File: packages/gatsby-source-shopify/README.md [hunk 2]
```diff
@@ -70,6 +70,29 @@ plugins: [
       // Possible values are: 'shop' and 'content'.
       // Defaults to ['shop', 'content'].
       includeCollections: ["shop", "content"],
+
+      // Allow overriding the default queries
+      // This allows you to include/exclude extra fields when sourcing nodes
+      // Available keys are: articles, blogs, collections, products, shopPolicies, and pages
+      // Queries need to accept arguments for first and after
+      // Defaults can be found in ./src/queries.js
```
Comment (pieh) [inline packages/gatsby-source-shopify/README.md:78]: Users can read this README on npm, gatsbyjs.org or github - `./src/queries.js` will only be helpful on github. Additionally that file doesn't contain the keys for `shopifyQueries` map, just query string values.

It almost feel like we need to include default queries so this README works as standalone. Not great solution but maybe we could have some appendix at the end of this file and link to it from here?

**Comment by DanielTamkin** [comment]
Hey! Looking to implement the [pagination solution suggested by Shopify](https://shopify.dev/concepts/graphql/pagination), thinking a custom query will be the best way to go about getting some of the keys i cannot usually get through this plugin. Can someone explain how to go about invoking a custom query for this plugin?

File: packages/gatsby-source-shopify/src/gatsby-node.js [hunk 3]
```diff
@@ -53,10 +53,22 @@ export const sourceNodes = async (
...
   const client = createClient(shopName, accessToken, apiVersion)
 
+  const defaultQueries = {
+    articles: ARTICLES_QUERY,
+    blogs: BLOGS_QUERY,
+    collections: COLLECTIONS_QUERY,
+    products: PRODUCTS_QUERY,
+    shopPolicies: SHOP_POLICIES_QUERY,
+    pages: PAGES_QUERY,
+  }
```
Comment (muescha) [inline packages/gatsby-source-shopify/src/gatsby-node.js:68]: can this refactored to export the map directly from `./queries` like in the readme?

**Comment by muescha** [comment]
@DanielTamkin 

can you please create a new issue ?

otherwise your question get lost in this already merged PR

File: packages/gatsby-source-shopify/src/gatsby-node.js (hunk 3 above)
Comment (muescha) [reply packages/gatsby-source-shopify/src/gatsby-node.js:68]: --> #25811 `refactor(gatsby-source-shopify): inline defaultqueries`

**Comment by abohannon** [comment]
I opened an issue for this. I'm also not able to query `availableQuantity` on variants nor `totalInventory` on products.

**Comment by nemethricsi** [comment]
I'm not sure how should I query it I did so in `gatsby-config.js`:

Enabled `unauthenticated_read_product_inventory`

According to the [docs](https://shopify.dev/docs/storefront-api/reference/products/productvariant) this field should be on `ProductVariant`

But it failes because it does not find `quantityAvailable` on `ProductVariantConnection`....

Do you have any clue? 

Thank you!

```graphql
shopifyQueries: {
  products: `
    query GetProducts($first: Int!, $after: String) {
      products(first: $first, after: $after) {
        pageInfo {
            hasNextPage
         }
         edges {
           cursor
              node {
                availableForSale
                variants {
                  quantityAvailable
                }
              }
            }
          }
        }
  `,
},
```

**Pull request merged**